bin/pyrdp-convert.py

```diff
@@ -0,0 +1,433 @@
+            self.dst = ip.dst
+            # Create the TLS session context.
+            self.tls = tlsSession(ipsrc=ip.src, ipdst=ip.dst, sport=tcp.sport, dport=tcp.dport, connection_end='server')
+
+        # Mirror the session if the packet is flowing in the opposite direction.
+        if self.tls.ipsrc != ip.src:
+            self.tls = self.tls.mirror()
+
+        try:
+            frame = TLS(p.load, tls_session=self.tls)
+        except AttributeError as e:
+            if not self.client:
+                return p  # ClientHelo is not sent yet: This is not TLS data.
+            else:
+                raise e  # Should be TLS data.
+
+        # Perform PDU reassembly.
+        if TLSApplicationData in frame:
+            payload = p.load
+            # There could be multiple nested TLS records within the
+            # same TCP packet. In that case we need to consume packets
+            # until The last layer is complete, otherwise it becomes
+            # extremely difficult to decrypt the rest of the stream.
+            tls = frame.lastlayer()
+            while tls.len - tls.deciphered_len - TLS_HDR_LEN > 0:
+                fragment = next(self.ipkt)
+
+                if Raw not in fragment:
+                    continue  # Skip TCP control.
+
+                payload += fragment.load
+                frame = TLS(payload, tls_session=self.tls)
+                tls = frame.lastlayer()
+
+        # FIXME: Maybe rebuild each TLSApplicationData to be a message entry in a single record?
+        tcp.remove_payload()
+        tcp.add_payload(frame)
+        self.tlsSession = frame.tls_session  # Update TLS Context.
+
+        # FIXME: Rather, check if the message is included in it to be sure?
+        msg = frame.msg[0]
+        if isinstance(msg, TLSClientHello):
+            self.client = pkcs_i2osp(msg.gmt_unix_time, 4) + msg.random_bytes
+        elif isinstance(msg, TLSServerHello):
+            self.server = pkcs_i2osp(msg.gmt_unix_time, 4) + msg.random_bytes
+
+        elif isinstance(msg, TLSNewSessionTicket):
+            # Session established, set master secret.
+            self.tls.rcs.derive_keys(client_random=self.client,
+                                    server_random=self.server,
+                                    master_secret=self.secret)
+
+            self.tls.wcs.derive_keys(client_random=self.client,
+                                    server_random=self.server,
+                                    master_secret=self.secret)
+        return p
+
+
+def decrypted(stream: PacketList, master_secret: bytes) -> Decrypted:
+    """An iterator function that decrypts a stream."""
+    return Decrypted(stream, master_secret)
+
+
+def getStreamInfo(s: PacketList) -> (str, str, float, bool):
+    """Attempt to retrieve an (src, dst, ts, isPlaintext) tuple for a data stream."""
+    packet = s[0]
+    if IP in packet:
+        # FIXME: Technically the IP traffic could be plaintext too.
```

Comment:
Currently, exported PDU traces do not contain an `IP` layer because they are all on a fake (storage) layer called EXPORT_PDU. So This code is a hacky way to check if it's an encrypted trace (has `IP` => TLS)

The reason I wrote this FIXME (I'll rephrase it now that you mentioned this) Is because it is theoretically possible (but unlikely) to have an unencrypted RDP connection, which would be a valid stream with IP and TCP layers but not have a TLS layer at all, making this code fail.